### PR TITLE
Remove unicorn reactions

### LINKED ISSUE #18277: Remove unicorn reactions
# Problem

The unicorn reactions, while being well-loved on DEV, can sometimes bring confusion to the users and it's not always on brand with all Forems. As we have discussed in several [Slack](https://forem-team.slack.com/archives/C03QCPEDDT3/p1658389672933249) [threads](https://forem-team.slack.com/archives/CAA200SH5/p1656305183488219?thread_ts=1655111706.931249&cid=CAA200SH5), we agree to remove the unicorn reactions for simplicity but think about how we can bring back the playfulness in a more customizable way at a later stage.

# To-dos

This is a list of things to do but needs some more pairs of eyes to check if these are exhaustive:
- Remove unicorn in all locations:
   - [x] Post page
   - [ ] Feed page
   - [ ] Notifications page (what do we show in place of the unicorn that happened in the past?)
   - [ ] `/dashboard/analytics` Reactions Summary :heart: 🦄 :bookmark: -> Also, there is a Unicorn legend in the graph
   - [ ] `/userID/some-post/stats` same as above
   - [ ] anywhere else?
- Handle legacy data
	- [ ] We need to recalculate all reactions and change how we show them

# Notes
1. As this involves lots of data recalculation, our overall strategy is that we will try to preserve old data for a while, instead of removing them for good all at once, in case there are scenarios we might want to consider using those data. 
2. This issue is a blocker to the following issues hence should be tackled first:
   - #18125 
   - #18127 

Product point person: @amywtlin.

### PR DESCRIPTION
**Pull request by joshpuetz**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This PR removes the unicorn reaction from articles. Users can not longer "unicorn" an article, nor will they see unicorn reactions in their statistic. Existing unicorn reactions are preserved, but not displayed. No summary data is recalculated, so reaction counts on articles with legacy unicorn reactions will be inaccurate.

## Related Tickets & Documents

- Closes #18277

## QA Instructions, Screenshots, Recordings

Just run locally

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [x] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [x] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

None

## [optional] What gif best describes this PR or how it makes you feel?

tumblr_nqow3o2EBO1u8yw5ao1_1280

### CODE REVIEW

File: app/javascript/analytics/dashboard.js [hunk 1]
```diff
@@ -250,20 +241,26 @@ function renderReferrers(data) {
...
 function removeCardElements() {
-  const el = document.getElementsByClassName("summary-stats")[0];
+  const el = document.getElementsByClassName('summary-stats')[0];
   el && el.remove();
 }
 
 function showErrorsOnCharts() {
-  const target = ['reactions-chart', 'comments-chart', 'followers-chart', 'readers-chart'];
-  target.forEach(id => {
+  const target = [
```
Comment (joshpuetz) [inline app/javascript/analytics/dashboard.js:249]: Some sort of autoformating kicked in here

File: lib/data_update_scripts/20220906190412_coerce_unicorn_reactions_to_likes.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
+module DataUpdateScripts
+  class CoerceUnicornReactionsToLikes
+    def run
+      # Find all unicorn reactions where there exists a like reaction on the same reactable
+      # by the same user, and delete them (These would cause errors when we try to flip all
+      # the unicorn reactions to a like)
+      ActiveRecord::Base.connection.execute(
+        <<-SQL,
+          DELETE FROM reactions
```
Comment (joshpuetz) [inline lib/data_update_scripts/20220906190412_coerce_unicorn_reactions_to_likes.rb:9]: Everything I've been able to find indicates this will be a row level lock in the database, instead of a table level lock. We should be deleting about 700k out of 9 million reaction records here

File: lib/data_update_scripts/20220906190412_coerce_unicorn_reactions_to_likes.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+                  AND unicorn_reactions.reactable_id = like_reactions.reactable_id
+                  AND unicorn_reactions.reactable_type = like_reactions.reactable_type
+              WHERE
+                unicorn_reactions.category = 'unicorn'
+                AND like_reactions.category = 'like')
+        SQL
+      )
+
+      # Find all reactions with unicorn reactions, flip them to like, (callbacks will update their scores)
+      Reaction.where(category: "unicorn").find_each(order: :desc) do |reaction|
```
Comment (joshpuetz) [inline lib/data_update_scripts/20220906190412_coerce_unicorn_reactions_to_likes.rb:23]: find_each will iterate through these reactions in batches of 1000, and the order: :desc is recommended in the data update script docs (although I wonder if it'll really slow down the query?). The general idea here is that after updating these unicorn reactions to like, we'll save them and utilize all of the callback code to recalculate their scores and reactions counts.

File: spec/lib/data_update_scripts/coerce_unicorn_reactions_to_likes_spec.rb [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  "lib/data_update_scripts/20220906190412_coerce_unicorn_reactions_to_likes.rb",
+)
+
+describe DataUpdateScripts::CoerceUnicornReactionsToLikes do
+  let(:user) { create(:user) }
+  let(:article) { create(:article) }
+  let(:unicorn_reaction) { create(:reaction, user: user, reactable: article, category: "like") }
+
+  before do
+    unicorn_reaction.update_attribute(:category, "unicorn") # This skips validations
```
Comment (joshpuetz) [inline spec/lib/data_update_scripts/coerce_unicorn_reactions_to_likes_spec.rb:12]: We have to play tricks to get a legacy unicorn reaction in the database since we've already removed it from the list of acceptable reactions 😉

**Comment by joshpuetz** [comment]
Setting to draft and marking as blocked as per conversation here: https://github.com/forem/forem/issues/18277#issuecomment-1239658500

TL;DR: we're pausing merging this PR until the replacement icon/widget for this space https://github.com/forem/forem/issues/18127, and will use a feature flag to toggle them.

**Comment by mirie** [comment]
@jaw6 can this one be closed?

**Comment by jaw6** [comment]
I think it's safe to close.
